README.txt
********************************************************************************
* Introduction                                                                 *
********************************************************************************

This plugin can be used to send periodic email reminders to bug reporters,
managers, and assignees.

The files in the `plugins/Reminder/scripts` directory should be run directly,
from the command line.

1. `bug_feedback_mail.php` sends emails to reporters listing all bugs awaiting
   their feedback.
2. `bug_reminder_mail.php` sends emails to assignees when bugs are approaching
   their due date.
3. `assigned_bugs.php` sends emails to assignees listing all open bugs that are
   assigned to them.

This plugin is build upon version 2.x.x of mantis and should be installed as
any other plugin. No Mantis scripts or tables are being altered.


********************************************************************************
* Installation                                                                 *
********************************************************************************

Like any other plugin.
After copying to your webserver :
- Start mantis ad administrator
- Select manage
- Select manage Plugins
- Select Install behind Reminder 2.06
- Once installed, click on the plugin-name for further configuration.

Make sure to have this statement in confg_inc.php:
$g_path          = 'http://path to your mantis installation/';


********************************************************************************
* Configuration options                                                        *
********************************************************************************

// What is the body of the E-mail
reminder_mail_subject	= "Following issue will be Due shortly";

// What is the subject of the grouped E-mail
reminder_group_subject	= "You have issues approaching their Due Date";

// What is the start of the body of the grouped E-mail
reminder_groupbody1	= "Please review the following issues";

// What is the end of the body of the grouped E-mail
reminder_groupbody2	= "Please do not reply to this message";

// perform for which project
reminder_project_id = '0'; means ALL
// you can define ultipleproject-id's separated by comma

// include or exclude project
reminder_include = ON; means Projects defined above are include

// how many days before Due date should we take into account
reminder_days_treshold  = 2;

// Should we use hours instead of days
reminder_hours		  	= OFF;

// Should we store this reminder as bugnote
reminder_store_as_note = OFF;
// only possible for handler

// For which status to send reminders
reminder_bug_status = ASSIGNED

// Ignore reminders for issues with no Due date set
reminder_ignore_unset = ON

// Ignore reminders for issues with Due dates in the past
reminder_ign_past = ON

// only valid for the mail function, downloads will always have duedates that have gone by

// Create overview per handler
reminder_handler = ON

// Group issues by Handler
reminder_group_issues = ON

// Group issues by project/handler
reminder_group_project = OFF

// Create overview per manager/project
reminder_manager_overview = ON
//
// access level for manager= 70
// this needs to be made flexible
// we will only produce overview for those projects that have a separate manager
//

// Select project to receive Feedback mail
reminder_feedback_project = 0; means ALL

// For which status to send feedbackreminders
reminder_bug_status = FEEDBACK


********************************************************************************
* Automatically generating mail                                                *
********************************************************************************

After this, bug_reminder_mail can be used via cron like this:
*/1440 *   *   *   * lynx --dump http://mantis.homepage.com/plugins/Reminder/scripts/bug_reminder_mail.php

or via command line interface
*/1440 *   *   *   * /usr/local/bin/php /path/to/mantis/plugins/Reminder/scripts/bug_reminder_mail.php

This line would send out a reminder every day.

You also can use a scheduled task under windows by calling a batch-file like:

REM *** this batch is running as a scheduled task under the ... Account ***
g:
cd \inetpub\wwwroot\mantis
php.exe plugins/Reminder/scripts/bug_reminder_mail.php


One can also schedule a job to prompt reporters to respond because their
issue has status Feedback. In that case replace bug_reminder_mail.php with
bug_feedback_mail.php


********************************************************************************
* Extras                                                                       *
********************************************************************************

On top of that, I have created a little variant which will create a
spreadsheet with issues getting due. Call script like:

http://www.YourMantisHome.com/plugins/Reminder/scripts/bug_due_overview2.php?days=2&status=50

If you do not apply parameters, the script will default to the above.
You can even poass on multiple statusses by separtaing them with a comma.

In the script directory you will also find a script called
bug_reminder_mail_test.php This script which you should call from within the
browser (once logged on) will provide useful feedback if things are not
working as expected. In case of a blank screen, all is processed normally.

For option Days/Hours, the script will fetch the plugin definition.


********************************************************************************
* License                                                                      *
********************************************************************************

This plugin is distributed under the same conditions as Mantis itself (GPL).


********************************************************************************
* Bug reporting                                                                *
********************************************************************************

Main issue: http://www.mantisbt.org/bugs/view.php?id=10153

Log new issues against the [Plugin - Reminder] project on mantisbt.org's tracker:
http://www.mantisbt.org/bugs/view_all_bug_page.php?project_id=7

This code is available on github: https://github.com/mantisbt-plugins/Reminder


********************************************************************************
* Thanks                                                                       *
********************************************************************************

Mark Ziegler	German translation	May 2010


********************************************************************************
* Greetings                                                                    *
********************************************************************************

Cas Nuy
cas@nuy.info
http://www.nuy.info


CHANGELOG.txt
Version |Date      |Name       |Message
----------------------------------------------------------------------------------------------------------
0.91     23-02-2009		Initial version

1.0	 16-03-2009		Bugfixes
				Added bonus program 

1.03 	 22-04-2009		Bugfixes
				Added program to request feedback from users

1.04	 23-04-2010		Bugfixes

1.05	 23-04-2010		Bugfixes
				Adjusted lay-out of emails

1.06	 26-04-2010		Bugfixes
				Adjusted authorisations when scheduled

1.07	 26-04-2010		Bugfixes
				Corrected the scheduling facilities

1.08	 27-04-2010		Bugfixes
				Added a test script to facilitate local verifying

1.09	 28-04-2010		Bugfixes
				Added results screen when ran from within browser

1.10	 29-04-2010		Bugfixes
				All fields translatable
				Compatible with 1.2.1

1.11	 07-05-2010		Added German translation ( thanks to Mark Ziegler)	
				Converted all mysql_query calls to mantis api calls

1.12	 08-02-2011		Protected 2 scripts to be ran from within the browser.
				bug_feedback_mail.php & bug_reminder_mail.php

1.20	 21-03-2011 cas		Added the option to use hours instead of days

1.21	 28-06-2012 johmart	Hours/days setting was mixed up.
                                Introduced Multiple Status selection
                                French translation aded

1.22     20-08-2012 funta222    Fixed Changelog, added missing 1.21 changes
                                Fixed broken multi status intoduced in 1.21
                                Fixed test scripts 
                                Feedback Reminder Email receipient now defaults to Reporter. Rename scripts\bug_feedback_mail_handler.php to get old handler bahavior                                      
                                Fixed Excel2010 warning abput wrong file typ: Csv vs. Xls
                                Introduced setting for character used as Csv column sparator (defaults to ; now)

xxx		2013-03-07	samwilson	Moved to Github for easier collaboration.
								Fixed paths in mailing scripts, so they can be called from anywhere.
								Added assigned_bugs.php mailing script, to email assignees of open bugs.
								Updated READMEs.
1.23	2016-05-14	cas			Added option to include/exclude multiple projects in selection
2.01	2017-01-21	cas			Compatible version for Mantis 2.0
2.02							Bugfixes
2.03							Bugfixes
2.04							Bugfixes
2.06	2018-08-28	Cas			Bugfixes, removed deprecated functions


scripts/bug_feedback_mail_test.php
<?php
# Make sure this script doesn't run via the webserver

# This page sends an E-mail to the REPORTER if an issue is awaiting feedback
# No real email is sent not are notes created for the various issues
#

require_once( '../../../core.php' );
$t_login	= config_get( 'plugin_Reminder_reminder_login' );
$ok=auth_attempt_script_login( $t_login );
$t_core_path = config_get( 'core_path' );
require_once( $t_core_path.'email_api.php' );
$t_bug_table	= db_get_table( 'bug' );
$t_project		= config_get('plugin_Reminder_reminder_feedback_project');
$status			= config_get('plugin_Reminder_reminder_feedback_status');
$t_rem_body1	= config_get( 'plugin_Reminder_reminder_group_body1' );
$t_rem_body2	= config_get( 'plugin_Reminder_reminder_group_body2' );
$query = "select id,reporter_id,handler_id,project_id from $t_bug_table where status in (".implode(",", $status).") ";
$t_rem_include	= config_get('plugin_Reminder_reminder_include');
$t_rem_projects	= "(";
$t_rem_projects	.= config_get('plugin_Reminder_reminder_project_id');
$t_rem_projects	.= ")";
if (ON==$t_rem_include){
	if ($t_rem_projects <>"0") {
		$query .= " and $t_bug_table.project_id IN ". $t_rem_projects;
	}
}else{
	$query .= " and $t_bug_table.project_id NOT IN ".$t_rem_projects;
}
$query .= " order by reporter_id";

$results = db_query( $query );
if ($results){
	$start = true ;
	$list= "";
	// first group and store feedback reminder per issue
	while ($row1 = db_fetch_array($results)) {
		$id 	   	= $row1['id'];
		$handler	= $row1['handler_id'];
		$project  = $row1['project_id'];
		$reporter = $row1['reporter_id'];

    echo "id: ".$id." handler_id: ".$handler." reporter_id:".$reporter." project_id: ".$project."\n<br>";

    if ($start){
			//$handler2 = $handler ;
			$reporter2 = $reporter ;
			$start = false ;
		}
		if ($reporter== $reporter2){
			$list .=" \n\n";
			$list .= string_get_bug_view_url_with_fqdn( $id, $reporter2 );
		} else {
			// now send the grouped email
			$body  = $t_rem_body1. " <br><br>\n\n";
			$body .= $list. " <br>\n\n";
			$body .= $t_rem_body2;
			$result = email_group_reminder( $reporter2, $body);
			$reporter2 = $reporter;
			$list= string_get_bug_view_url_with_fqdn( $id, $reporter2 );
		}
		$list .=" <br>\n";
	}
	// handle last grouped email
	if ($results){
		$body  = $t_rem_body1. " \n\n";
		$body .= $list. " \n\n";
		$body .= $t_rem_body2;
		$result = email_group_reminder( $reporter2, $body);
	} else {
  		echo 'Query-Manager had no results '.$query ;
	  	echo '<br>';
	}

}
if (php_sapi_name() !== 'cli'){
	echo config_get( 'plugin_Reminder_reminder_finished' );
}

# Send Grouped reminder
function email_group_reminder( $p_user_id, $issues ) {
	$t_username = user_get_field( $p_user_id, 'username' );
	$t_email = user_get_email( $p_user_id );
	$t_message = $issues ;
	$t_subject	= config_get( 'plugin_Reminder_reminder_subject' );

  if( !is_blank( $t_email ) ) {

	echo '------------------------------- <br>';
	echo $t_email;
		echo '**';
		echo $t_subject;
		echo '<br>';
		echo $t_message;
		echo '<br>';
		echo '------------------------------- <br>';

		// email_store( $t_email, $t_subject, $t_message );
		if( OFF == config_get( 'email_send_using_cronjob' ) ) {
				// email_send_all();
		}
	}
}


scripts/bug_reminder_mail_test.php
<?php
# This page tests sending an E-mail if a due date is getting near
# No real email is sent nor are notes created for the various issues
require_once( '../../../core.php' );
$t_login	= config_get( 'plugin_Reminder_reminder_login' );
$ok=auth_attempt_script_login( $t_login );
$t_core_path = config_get( 'core_path' );
require_once( $t_core_path.'bug_api.php' );
require_once( $t_core_path.'email_api.php' );
require_once( $t_core_path.'bugnote_api.php' );

$t_bug_table	= db_get_table( 'bug' );
$t_man_table	= db_get_table( 'project_user_list' );
$t_rem_days		= config_get( 'plugin_Reminder_reminder_days_treshold' );
$t_rem_status	= config_get( 'plugin_Reminder_reminder_bug_status' );
$t_rem_body		= config_get( 'plugin_Reminder_reminder_mail_subject' );
$t_rem_store	= config_get( 'plugin_Reminder_reminder_store_as_note' );
$t_rem_ignore	= config_get( 'plugin_Reminder_reminder_ignore_unset' );
$t_rem_ign_past	= config_get( 'plugin_Reminder_reminder_ignore_past' );
$t_rem_handler	= config_get( 'plugin_Reminder_reminder_handler' );
$t_rem_group1	= config_get( 'plugin_Reminder_reminder_group_issues' );
$t_rem_group2	= config_get( 'plugin_Reminder_reminder_group_project' );
$t_rem_manager	= config_get( 'plugin_Reminder_reminder_manager_overview' );
$t_rem_subject	= config_get( 'plugin_Reminder_reminder_group_subject' );
$t_rem_body1	= config_get( 'plugin_Reminder_reminder_group_body1' );
$t_rem_body2	= config_get( 'plugin_Reminder_reminder_group_body2' );

$t_rem_hours	= config_get('plugin_Reminder_reminder_hours');
if (ON != $t_rem_hours){
	$multiply=24;
} else{
	$multiply=1;
}
//
// access level for manager= 70
// this needs to be made flexible
// we will only produce overview for those projects that have a separate manager
//
$baseline	= time(true)+ ($t_rem_days*$multiply*60*60);
$basenow	= time(true);

echo "Path setting retrieved : ".config_get('path');
echo "<br>";

if ( ON == $t_rem_handler ) {
	echo 'Query-handler being executed' ;
	echo '<br>';
	$query = "select id,handler_id,project_id from $t_bug_table where status in (".implode(",", $t_rem_status).") and due_date<=$baseline and handler_id>0 ";
	if ( ON == $t_rem_ign_past ) {
			$query .=" and due_date>=$basenow" ;
	} else{
		if ( ON == $t_rem_ignore ) {
			$query .=" and due_date>1" ;
		}
	}
$t_rem_include	= config_get('plugin_Reminder_reminder_include');
$t_rem_projects	= "(";
$t_rem_projects	.= config_get('plugin_Reminder_reminder_project_id');
$t_rem_projects	.= ")";
if (ON==$t_rem_include){
	if ($t_rem_projects <>"0") {
		$query .= " and $t_bug_table.project_id IN ". $t_rem_projects;
	}
}else{
	$query .= " and $t_bug_table.project_id NOT IN ".$t_rem_projects;
}
	
	if ( ON == $t_rem_group1 ) {
		$query .=" order by handler_id" ;
	}else{
		if ( ON == $t_rem_group2 ) {
			$query .=" order by $t_bug_table.project_id,handler_id" ;
		}
	}
	echo $query;
	echo "<br>"  ;
	$results = db_query( $query );
	$resnum=db_num_rows($results);
	echo $resnum;
	echo "<br>"  ;
	if ( OFF == $t_rem_group1 ) {
		if ($results) {
			while ($row1 = db_fetch_array($results)) {
				$id 		= $row1['id'];
				$handler	= $row1['handler_id'];
				echo $id;
				echo '*';
				echo $handler;
				echo "<br>";
				$list = string_get_bug_view_url_with_fqdn( $id, $handler2 );
				$body  = $t_rem_body1;
				$body .= "<br>";
				$body .= $list;
				$body .= "<br>";
				$body .= $t_rem_body2;
				$result = email_group_reminder( $handler, $body );
				# Add reminder as bugnote if store reminders option is ON.
				if ( ON == $t_rem_store ) {
					$t_attr = '|'.$handler2.'|';
//					bugnote_add( $id, $t_rem_body, 0, config_get( 'default_reminder_view_status' ) == VS_PRIVATE, REMINDER, $t_attr, NULL, FALSE );
				}
			}
		} else {
			echo 'Query-handler had no results'.$query ;
			echo '<br>';
		}
	} else {
		if ($results){
			$start = true ;
			$list= "";
			// first group and store reminder per issue
			while ($row1 = db_fetch_array($results)) {
				$id 		= $row1['id'];
				$handler	= $row1['handler_id'];
				$project	= $row1['project_id'];
				echo $id;
				echo '*';
				echo $handler;
				echo '*';
				echo $project;
				echo "<br>";
				if ($start){
					$handler2 = $handler ;
					$start = false ;
				}
				if ($handler==$handler2){
					$list .="<br>";
					$list .= string_get_bug_view_url_with_fqdn( $id, $handler2 );
					# Add reminder as bugnote if store reminders option is ON.
					if ( ON == $t_rem_store ) {
						$t_attr = '|'.$handler2.'|';
//						bugnote_add( $id, $t_rem_body, 0, config_get( 'default_reminder_view_status' ) == VS_PRIVATE, REMINDER, $t_attr, NULL, FALSE );
					}
				} else {
					// now send the grouped email
					$body  = $t_rem_body1;
					$body .= "<br>";
					$body .= $list;
					$body .= "<br>";
					$body .= $t_rem_body2;
					$result = email_group_reminder( $handler2, $body);
					$handler2 = $handler ;
					$list ="<br>";
					$list= string_get_bug_view_url_with_fqdn( $id, $handler2 );
					# Add reminder as bugnote if store reminders option is ON.
					if ( ON == $t_rem_store ) {
						$t_attr = '|'.$handler2.'|';
//						bugnote_add( $id, $t_rem_body, 0, config_get( 'default_reminder_view_status' ) == VS_PRIVATE, REMINDER, $t_attr, NULL, FALSE );
					}
				}
			}
			// handle last one
			if ($resnum>0){
				// now send the grouped email
				$body  = $t_rem_body1;
				$body .= "<br>";
				$body .= $list;
				$body .= "<br>";
				$body .= $t_rem_body2;
				$result = email_group_reminder( $handler2, $body);

			} else{
				echo 'Query-Handler had no results '.$query ;
				echo '<br>';
			}
			//
		}else {
			echo 'Query-handler had no results '.$query ;
			echo '<br>';
		}
	}
}

if ( ON == $t_rem_manager ) {
	echo 'Query-Manager being executed' ;
	echo '<br>';
	// select relevant issues in combination with an assigned manager to the project
	$query  = "select id,handler_id,user_id from $t_bug_table,$t_man_table where status in (".implode(",", $t_rem_status).") and due_date<=$baseline ";
	if ( ON == $t_rem_ign_past ) {
			$query .=" and due_date>=$basenow" ;
	} else{
		if ( ON == $t_rem_ignore ) {
			$query .=" and due_date>1" ;
		}
	}
$t_rem_include	= config_get('plugin_Reminder_reminder_include');
$t_rem_projects	= "(";
$t_rem_projects	.= config_get('plugin_Reminder_reminder_project_id');
$t_rem_projects	.= ")";
if (ON==$t_rem_include){
	if ($t_rem_projects <>"0") {
		$query .= " and $t_bug_table.project_id IN ". $t_rem_projects;
	}
}else{
	$query .= " and $t_bug_table.project_id NOT IN ".$t_rem_projects;
}

	$query .=" and $t_bug_table.project_id=$t_man_table.project_id and $t_man_table.access_level=70" ;
	$query .=" order by $t_man_table.project_id,$t_man_table.user_id" ;
	

	
	echo $query;
	echo "<br>"  ;
	$results = db_query( $query );
	$resnum=db_num_rows($results);
	echo $resnum;
	echo "<br>"  ;
	if ($results){
		$start = true ;
		$list= "";
		// first group and store reminder per issue
		while ($row1 = db_fetch_array($results)) {
				$id 		= $row1['id'];
				$handler	= $row1['handler_id'];
				$manager	= $row1['user_id'];
			echo $id;
			echo '*';
			echo $handler;
			echo '*';
			echo $manager;
			echo "<br>";
			if ($start){
				$man2 = $manager ;
				$start = false ;
			}
			if ($manager==$man2){
				$list .=" \n\n";
				$list .= string_get_bug_view_url_with_fqdn( $id, $man2 );
			} else {
				// now send the grouped email
				$body  = $t_rem_body1. " \n\n";
				$body .= $list. " \n\n";
				$body .= $t_rem_body2;
				$result = email_group_reminder( $man2, $body);
				$man2 = $manager ;
				$list= string_get_bug_view_url_with_fqdn( $id, $man2 );
				$list .= " \n\n";
			}
		}
		// handle last one
		if ($resnum>0){
			// now send the grouped email
			$body  = $t_rem_body1. " \n\n";
			$body .= $list. " \n\n";
			$body .= $t_rem_body2;
			$result = email_group_reminder( $man2, $body);

		}else{
				echo 'Query-Manager had no results '.$query ;
				echo '<br>';
			}
		//
	} else {
		echo 'Query-Manager had no results '.$query ;
		echo '<br>';
	}
}
echo '<br><br><br>';
echo 'Finished Reminder Test ';

# Send Grouped reminder
function email_group_reminder( $p_user_id, $issues ) {
	$t_username = user_get_field( $p_user_id, 'username' );
	$t_email = user_get_email( $p_user_id );
	$t_subject = config_get( 'plugin_Reminder_reminder_group_subject' );
	$t_message = $issues ;
	if( !is_blank( $t_email ) ) {
		echo $t_email;
		echo '**';
		echo $t_subject;
		echo '<br>';
		echo $t_message;
		echo '<br>';
//		email_store( $t_email, $t_subject, $t_message );
		if( OFF == config_get( 'email_send_using_cronjob' ) ) {
//			email_send_all();
		}
	}
}


Reminder.php
<?php
class ReminderPlugin extends MantisPlugin {

	function register() {
		$this->name        = 'Reminder';
		$this->description = lang_get( 'reminder_plugin_desc' );
		$this->version     = '2.06';
		$this->requires    = array('MantisCore'       => '2.0.0',);
		$this->author      = 'Cas Nuy';
		$this->contact     = 'Cas-at-nuy.info';
		$this->url         = 'http://www.nuy.info';
		$this->page        = 'config';
	}

	/*** Default plugin configuration.	 */
	function config() {
		return array(
			'reminder_mail_subject'			=> 'Following issue will be Due shortly' ,
			'reminder_days_treshold'		=> 2,
			'reminder_store_as_note'		=> OFF,
			'reminder_sender'				=> 'admin@example.com',
			'reminder_bug_status'			=> array(ASSIGNED),
			'reminder_ignore_unset'			=> ON,
			'reminder_ignore_past'			=> ON,
			'reminder_handler'				=> ON,
			'reminder_group_issues'			=> ON,
			'reminder_group_project'		=> OFF,
			'reminder_manager_overview'		=> ON,
			'reminder_group_subject'		=> "You have issues approaching their Due Date",
			'reminder_group_body1'			=> "Please review the following issues",
			'reminder_group_body2'			=> "Please do not reply to this message",
			'reminder_project_id'			=> "0",
			'reminder_include'				=>	ON,
			'reminder_login'				=> 'admin',
			'reminder_feedback_status'		=> array(FEEDBACK),
			'reminder_subject'				=> 'Issues requiring your attention',
			'reminder_finished'				=> 'Finished processing your selection',
			'reminder_hours'				=> OFF,
			'reminder_colsep'				=> ';',
			);
	}

	function init() {
		plugin_event_hook( 'EVENT_MENU_MANAGE', 'remdown' );
	}

	function remdown() {
		return array('<a href="'. plugin_page( 'bug_due_overview.php' ) . '">' . lang_get( 'reminder_download' ) . '</a>' );
	}

}


scripts/bug_reminder_mail.php
<?php
# Make sure this script doesn't run via the webserver
/*
if( php_sapi_name() != 'cli' ) {
	echo "It is not allowed to run this script with PHP SAPI name: " . php_sapi_name() . ".\n";
	exit( 1 );
}
*/

# This page sends an E-mail if a due date is getting near
# includes all due_dates not met
require_once( dirname( dirname( dirname( dirname( __FILE__ ) ) ) ) . DIRECTORY_SEPARATOR . 'core.php' );
$t_login	= config_get( 'plugin_Reminder_reminder_login' );
$ok=auth_attempt_script_login( $t_login );
$t_core_path = config_get( 'core_path' );
require_once( $t_core_path.'bug_api.php' );
require_once( $t_core_path.'email_api.php' );
require_once( $t_core_path.'bugnote_api.php' );
require_once( $t_core_path.'category_api.php' );
require_once( $t_core_path.'helper_api.php' );

$allok= true ;
$t_bug_table	= db_get_table( 'bug' );
$t_bug_text_table = db_get_table( 'bug_text' );
$t_man_table	= db_get_table( 'project_user_list' );

$t_rem_project	= config_get( 'plugin_Reminder_reminder_project_id' );
$t_rem_days		= config_get( 'plugin_Reminder_reminder_days_treshold' );
$t_rem_status	= config_get( 'plugin_Reminder_reminder_bug_status' );
$t_rem_body		= config_get( 'plugin_Reminder_reminder_mail_subject' );
$t_rem_store	= config_get( 'plugin_Reminder_reminder_store_as_note' );
$t_rem_ignore	= config_get( 'plugin_Reminder_reminder_ignore_unset' );
$t_rem_ign_past	= config_get( 'plugin_Reminder_reminder_ignore_past' );
$t_rem_handler 	= config_get( 'plugin_Reminder_reminder_handler' );
$t_rem_group1	= config_get( 'plugin_Reminder_reminder_group_issues' );
$t_rem_group2	= config_get( 'plugin_Reminder_reminder_group_project' );
$t_rem_manager	= config_get( 'plugin_Reminder_reminder_manager_overview' );
$t_rem_subject	= config_get( 'plugin_Reminder_reminder_group_subject' );
$t_rem_body1	= config_get( 'plugin_Reminder_reminder_group_body1' );
$t_rem_body2	= config_get( 'plugin_Reminder_reminder_group_body2' );

$t_rem_hours	= config_get('plugin_Reminder_reminder_hours');
if (ON != $t_rem_hours){
	$multiply=24;
} else{
	$multiply=1;
}

//
// access level for manager= 70
// this needs to be made flexible
// we will only produce overview for those projects that have a separate manager
//
$baseline	= time(true)+ ($t_rem_days*$multiply*60*60);
$basenow	= time(true);
if ( ON == $t_rem_handler ) {
	$query = "select bugs.id, bugs.handler_id, bugs.project_id, bugs.priority, bugs.category_id, bugs.status, bugs.severity, bugs.summary from $t_bug_table bugs JOIN $t_bug_text_table text ON (bugs.bug_text_id = text.id) where status in (".implode(",", $t_rem_status).") and due_date<=$baseline and handler_id<>0 ";
	if ( ON == $t_rem_ign_past ) {
			$query .=" and due_date>=$basenow" ;
	} else{
		if ( ON == $t_rem_ignore ) {
			$query .=" and due_date>1" ;
		}
	}
	
	$t_rem_include	= config_get('plugin_Reminder_reminder_include');
$t_rem_projects	= "(";
$t_rem_projects	.= config_get('plugin_Reminder_reminder_project_id');
$t_rem_projects	.= ")";
if (ON==$t_rem_include){
	if ($t_rem_projects <> 0) {
		$query .= " and bugs.project_id IN ". $t_rem_projects;
	}
}else{
	$query .= " and bugs.project_id NOT IN ".$t_rem_projects;
}
	
	if ( ON == $t_rem_group1 ) {
		$query .=" order by handler_id" ;
	}else{
		if ( ON == $t_rem_group2 ) {
			$query .=" order by bugs.project_id,handler_id" ;
		}
	}
	$results = db_query( $query );
	$resnum=db_num_rows($results);
	if ( OFF == $t_rem_group1 ) {
		if ($results) {
			while ($row1 = db_fetch_array($results)) {
				$id 		= $row1['id'];
				$handler	= $row1['handler_id'];
				$list = string_get_bug_view_url_with_fqdn( $id, $handler2 );
				$body  = $t_rem_body1. " \n\n";
				$body .= $list. " \n\n";
				$body .= $t_rem_body2;
				$result = email_group_reminder( $handler, $body );
				# Add reminder as bugnote if store reminders option is ON.
				if ( ON == $t_rem_store ) {
					$t_attr = '|'.$handler2.'|';
					bugnote_add( $id, $t_rem_body, 0, config_get( 'default_reminder_view_status' ) == VS_PRIVATE, REMINDER, $t_attr, NULL, FALSE );
				}
			}
		}
	} else {
		if ($results){
			$start = true ;
			$list= "";
			// first group and store reminder per issue
			while ($row1 = db_fetch_array($results)) {
				$id 		= $row1['id'];
				$handler	= $row1['handler_id'];
				$project	= $row1['project_id'];
				if ($start){
					$handler2 = $handler ;
					$start = false ;
				}
				if ($handler==$handler2){
					$list .= formatBugEntry($row1);
					# Add reminder as bugnote if store reminders option is ON.
					if ( ON == $t_rem_store ) {
						$t_attr = '|'.$handler2.'|';
						bugnote_add( $id, $t_rem_body, 0, config_get( 'default_reminder_view_status' ) == VS_PRIVATE, REMINDER, $t_attr, NULL, FALSE );
					}
				} else {
					// now send the grouped email
					$body  = $t_rem_body1. " \n\n";
					$body .= $list. " \n\n";
					$body .= $t_rem_body2;
					$result = email_group_reminder( $handler2, $body);
					$handler2 = $handler ;
					$list = formatBugEntry($row1);
					# Add reminder as bugnote if store reminders option is ON.
					if ( ON == $t_rem_store ) {
						$t_attr = '|'.$handler2.'|';
						bugnote_add( $id, $t_rem_body, 0, config_get( 'default_reminder_view_status' ) == VS_PRIVATE, REMINDER, $t_attr, NULL, FALSE );
					}
				}
			}
			// handle last one
			if ($resnum>0){
				// now send the grouped email
				$body  = $t_rem_body1. " \n\n";
				$body .= $list. " \n\n";
				$body .= $t_rem_body2;
				$result = email_group_reminder( $handler2, $body);

			}
			//
		}
	}
}

if ( ON == $t_rem_manager ) {
	// select relevant issues in combination with an assigned manager to the project
	$query  = "select id,handler_id,user_id from $t_bug_table,$t_man_table where status in (".implode(",", $t_rem_status).") and due_date<=$baseline ";
	if ( ON == $t_rem_ign_past ) {
			$query .=" and due_date>=$basenow" ;
	} else{
		if ( ON == $t_rem_ignore ) {
			$query .=" and due_date>1" ;
		}
	}
//	if ( $t_rem_project>0 ) {
//		$query .=" and $t_bug_table.project_id=$t_rem_project" ;
//	}

$t_rem_include	= config_get('plugin_Reminder_reminder_include');
$t_rem_projects	= "(";
$t_rem_projects	.= config_get('plugin_Reminder_reminder_project_id');
$t_rem_projects	.= ")";
if (ON==$t_rem_include){
	if ($t_rem_projects <> 0) {
		$query .= " and project_id IN ". $t_rem_projects;
	}
}else{
	$query .= " and project_id NOT IN ".$t_rem_projects;
}
	$query .=" and $t_bug_table.project_id=$t_man_table.project_id and $t_man_table.access_level=70" ;
	$query .=" order by $t_man_table.project_id,$t_man_table.user_id" ;
	$results = db_query( $query );
	$resnum=db_num_rows($results);
	if ($results){
		$start = true ;
		$list= "";
		// first group and store reminder per issue
		while ($row1 = db_fetch_array($results)) {
			$id 		= $row1['id'];
			$handler	= $row1['handler_id'];
			$manager	= $row1['user_id'];
			if ($start){
				$man2 = $manager ;
				$start = false ;
			}
			if ($manager==$man2){
				$list .=" \n\n";
				$list .= string_get_bug_view_url_with_fqdn( $id, $man2 );
			} else {
				// now send the grouped email
				$body  = $t_rem_body1. " \n\n";
				$body .= $list. " \n\n";
				$body .= $t_rem_body2;
				$result = email_group_reminder( $man2, $body);
				$man2 = $manager ;
				$list= string_get_bug_view_url_with_fqdn( $id, $man2 );
				$list .= " \n\n";
			}
		}
		// handle last one
		if ($resnum>0){
			// now send the grouped email
			$body  = $t_rem_body1. " \n\n";
			$body .= $list. " \n\n";
			$body .= $t_rem_body2;
			$result = email_group_reminder( $man2, $body);

		}
		//
	}
}
if (php_sapi_name() !== 'cli'){
	echo config_get( 'plugin_Reminder_reminder_finished' );
}

# Send Grouped reminder
function email_group_reminder( $p_user_id, $issues ) {
	$t_username = user_get_field( $p_user_id, 'username' );
	$t_email = user_get_email( $p_user_id );
	$t_subject = config_get( 'plugin_Reminder_reminder_group_subject' );
	$t_message = $issues ;
	if( !is_blank( $t_email ) ) {
		email_store( $t_email, $t_subject, $t_message );
		if( OFF == config_get( 'email_send_using_cronjob' ) ) {
			email_send_all();
		}
	}
}

function formatBugEntry($data){
	lang_push( user_pref_get_language( $data['handler_id'] ) );

	$p_visible_bug_data = $data;
	$p_visible_bug_data['email_project'] = project_get_name( $data['project_id']);
	$p_visible_bug_data['email_category'] = category_get_name($data['category_id']);

	$t_email_separator1 = config_get( 'email_separator1' );
	$t_email_separator2 = config_get( 'email_separator2' );

	$p_visible_bug_data['email_bug'] = $data['id'];
	$p_visible_bug_data['email_status'] = get_enum_element( 'status', $p_visible_bug_data['status'], $data['handler_id'], $data['project_id'] );
	$p_visible_bug_data['email_severity'] = get_enum_element( 'severity', $p_visible_bug_data['severity'] );
	$p_visible_bug_data['email_priority'] = get_enum_element( 'priority', $p_visible_bug_data['priority'] );
	$p_visible_bug_data['email_reproducibility'] = get_enum_element( 'reproducibility', $p_visible_bug_data['reproducibility'] );
	$p_visible_bug_data['email_summary'] = $data['summary'];

	$t_message = $t_email_separator1 . " \n";
	$t_message .= string_get_bug_view_url_with_fqdn( $data['id'], $data['handler_id'] ) . " \n";
	$t_message .= $t_email_separator1 . " \n";

	$t_message .= email_format_attribute( $p_visible_bug_data, 'email_project' );
	$t_message .= email_format_attribute( $p_visible_bug_data, 'email_bug' );
	$t_message .= email_format_attribute( $p_visible_bug_data, 'email_category' );
	$t_message .= email_format_attribute( $p_visible_bug_data, 'email_priority' );
	$t_message .= email_format_attribute( $p_visible_bug_data, 'email_status' );
	$t_message .= $t_email_separator1 . " \n";

	$t_message .= email_format_attribute( $p_visible_bug_data, 'email_summary' );
	$t_message .= $t_email_separator1 . " \n\n\n";

	return $t_message;
}
